File: cpp/leetcode/446.cpp
class Solution {
public:
    int numberOfArithmeticSlices(vector<int>& A) {
        if (A.empty()) return 0;
        int n = A.size();
        vector<unordered_map<long long, int >> dp(n);
        unordered_set<int> s(A.begin(), A.end());
        int res = 0;
        for (int i = 1; i < n; ++i) {
            for (int j = i-1; j >= 0; --j) {
                long long d = (long long)A[i] - (long long)A[j];
                int tmp = dp[j].count(d) ? dp[j][d] : 0;
                if (tmp) res += tmp;
                if (s.count(A[i]+d)) dp[i][d] += 1 + tmp;
            }
        }
        return res;
    }
};

File: python/leetcode/446.py
class Solution(object):
    def numberOfArithmeticSlices(self, A):
        dp = [defaultdict(int) for i in range(len(A))]
        res = 0
        for i in range(1, len(A)):
            for j in range(i):
                step = A[i]-A[j]
                dp[i][step] += 1
                if step in dp[j]:
                    dp[i][step]+= dp[j][step]
                    res += dp[j][step]
        return res